xm-commons-permission/src/main/java/com/icthh/xm/commons/permission/service/PermissionCheckService.java

```diff
@@ -146,124 +156,210 @@ private boolean checkPermission(Authentication authentication,
+            .collect(toList());
+
+        if (!validCondition && reactionStrategies.contains(ReactionStrategy.SKIP)) {
+            throw new SkipPermissionException(
+                "Skip permission",
+                permissions.stream()
+                    .map(permission -> permission.getRoleKey() + ":" + permission.getPrivilegeKey())
+                    .collect(toList())
+            );
         } else if (!validCondition) {
             return false;
         }
-        log(logPermission, Level.INFO,
+
+        log(logPermission,
+            Level.INFO,
             "access granted: privilege={}, role={}, userKey={}",
-            privilegeKey, roleKey, getUserKey());
+            privilegeKey,
+            roleKey,
+            getUserKey()
+        );
         return true;
     }
 
     private boolean checkRole(Authentication authentication, Object privilege, boolean logPermission) {
-        String roleKey = getRoleKey(authentication);
+        Collection<String> roleKeys = getRoleKeys(authentication);
 
-        if (RoleConstant.SUPER_ADMIN.equals(roleKey)) {
-            log(logPermission, Level.INFO,
+        if (roleKeys.contains(SUPER_ADMIN)) {
+            log(logPermission,
+                Level.INFO,
                 "access granted: privilege={}, role=SUPER-ADMIN, userKey={}",
-                privilege, getUserKey());
+                privilege,
+                getUserKey()
+            );
             return true;
         }
 
-        if (!roleService.getRoles(TenantContextUtils.getRequiredTenantKeyValue(tenantContextHolder.getContext()))
-            .containsKey(roleKey)) {
-            log(logPermission, Level.ERROR,
+        if (listsNotEqualsIgnoreOrder(
+            roleService.getRoles(getRequiredTenantKeyValue(tenantContextHolder.getContext())).keySet(),
+            roleKeys
+        )) {
+            log(
+                logPermission,
+                Level.ERROR,
                 "access denied: privilege={}, role={}, userKey={} due to role is missing",
-                privilege, roleKey, getUserKey());
+                privilege,
+                roleKeys,
+                getUserKey()
+            );
             throw new AccessDeniedException("Access is denied");
         }
 
         return false;
     }
 
-    private static boolean isConditionValid(Expression expression, StandardEvaluationContext context) {
-        boolean result;
-        if (expression == null || StringUtils.isEmpty(expression.getExpressionString())) {
-            result = true;
-        } else {
-            try {
-                result = expression.getValue(context, Boolean.class);
-            } catch (Exception e) {
-                result = false;
-                log.error("Exception while getting value ", e);
-            }
-        }
-        return result;
+    private boolean isEnvConditionNotValid(Collection<Permission> permissions, StandardEvaluationContext context) {
```

Comment:
think isEnvConditionNotValid() and isResourceConditionNotValid() might be combined into one method. The difference only in one line